Documentation/Multiple Selections.md

```diff
@@ -0,0 +1,96 @@
+### Multiple Carets / Multiple Cursors / Multiple Selections
+
+If the host supports multiple selections, VsVim will also support them.
+
+#### Architecture
+
+Most parts of VsVim are unware of multiple selections. In order to keep things
+simple and preserve the single selection model for most of VsVim, there are two
+major components that handle multiple selections:
+
+- Running mode commands for all selections (`ICommonOperations.RunForAllSelections`)
+- The multi-selection tracker (`MultiSelectionTracker`)
+
+For commands where it makes sense to run them for each caret or selection, the
+corresponding command runner will wrap the running of a single command into a
+batch execution for each selection using `RunForAllSelections`. Then, while
+running an individual command, the multi-selection infrastructure will:
+
+- Set a single temporary selection, including a caret position
+- Run the command as if there were a single selection
+- Determine the desired resulting selection from the command result
+
+When all the individual commands have run, the accumulated selections will be
+set all at once. This causes the non-multi-selection aware code to still see
+the primary selection as the real and only selection, while maintaining the
+secondary selections behind the scenes.
+
+#### The Multi-Selection API
+
+Unfortunately, not all versions of the editor API support multiple selections.
+In order to handle this, there are two API entry points in the Vim host:
+
+- `IVimHost.GetSelectedSpans` - member
+- `IVimHost.SetSelectedSpans` - member

```

Comment:
Once we move past VS2015 support will this API dissapear? At that point we should have direct access to the multi-selection APIs in VimCore hence this shouldn't be necessary anymore (to my understanding).

Reply:
Frankly, I didn't fully understand how Vim.Specific assembly worked and the one tool in my toolbox for escaping from editor-land to WPF-land was `IVimHost` so that's what I initially used. I gradually realized how it was working and that I could eliminate duplicate code by using Vim.Specific. At that point, I was just one step away from ripping out the `IVimHost` API and accessing the `ISelectionUtilService` (which is already in F#) directly from common operations.

I will take a crack at that change later tonight.